src/foam/nanos/ruler/RuleEngine.java

```diff
@@ -51,10 +53,76 @@ public void setDelegate(DAO delegate) {
    * @param oldObj - Old FObject supplied to rules for execution
    */
   public void execute(List<Rule> rules, FObject obj, FObject oldObj) {
-    applyRules(rules, obj, oldObj);
+    CompoundContextAgency compoundAgency = new CompoundContextAgency();
+    ContextualizingAgency agency = new ContextualizingAgency(compoundAgency, userX_, getX());
+    for (Rule rule : rules) {
+      if ( stops_.get() ) break;
+      if ( ! isRuleApplicable(rule, obj, oldObj)) continue;
+      PM pm = new PM();
+      pm.setClassType(RulerDAO.getOwnClassInfo());
+      pm.setName(rule.getName());
```

Comment:
Do we know the daoKey? Maybe it would be better to do something like:
 pm.setName(daoKey + ":" + rule.getName());

Or, does the rule know the daoKey, so it would be rule.daoKey(_) + ":" + rule.getName()?